File: src/lib/routing/trailAccess.ts
/**
 * Trailhead orientation — gravity uses elevation (top = start), else nearest end.
 * Mirrors `mobile/lib/domain/routing/trail_access.dart`.
 */

export const kTrailElevDecideMinM = 8;

export type OrientedTrail = {
  geometry: [number, number][];
  entry: [number, number];
  exit: [number, number];
  reversed: boolean;
  usedElevation: boolean;
};

export function trailAccessHaversineKm(
  lat1: number,
  lng1: number,
  lat2: number,
  lng2: number
): number {
  const r = 6371;
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLng = ((lng2 - lng1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos((lat1 * Math.PI) / 180) *
      Math.cos((lat2 * Math.PI) / 180) *
      Math.sin(dLng / 2) ** 2;
  return 2 * r * Math.asin(Math.min(1, Math.sqrt(a)));
}

export function orientTrail(opts: {
  geometry: [number, number][];
  fromLng: number;
  fromLat: number;
  startElevM?: number | null;
  endElevM?: number | null;
  preferDownhill: boolean;
}): OrientedTrail {
  const { geometry, fromLng, fromLat, startElevM, endElevM, preferDownhill } =
    opts;
  if (geometry.length < 2) {
    const p = geometry[0] ?? [fromLng, fromLat];
    return {
      geometry,
      entry: p,
      exit: p,
      reversed: false,
      usedElevation: false,
    };
  }

  const first = geometry[0];
  const last = geometry[geometry.length - 1];
  let reverse = false;
  let usedElevation = false;

  if (
    preferDownhill &&
    startElevM != null &&
    endElevM != null &&
    Number.isFinite(startElevM) &&
    Number.isFinite(endElevM) &&
    Math.abs(startElevM - endElevM) >= kTrailElevDecideMinM
  ) {
    usedElevation = true;
    reverse = startElevM < endElevM;
  } else {
    const dFirst = trailAccessHaversineKm(fromLat, fromLng, first[1], first[0]);
    const dLast = trailAccessHaversineKm(fromLat, fromLng, last[1], last[0]);
    reverse = dLast < dFirst;
  }

  const geo = reverse ? [...geometry].reverse() : geometry;
  return {
    geometry: geo,
    entry: geo[0],
    exit: geo[geo.length - 1],
    reversed: reverse,
    usedElevation,
  };
}

export async function fetchEndpointElevations(
  a: [number, number],
  b: [number, number]
): Promise<{ startM: number | null; endM: number | null }> {
  try {
    const res = await fetch("/api/elevation", {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({
        locations: [
          { lat: a[1], lng: a[0] },
          { lat: b[1], lng: b[0] },
        ],
      }),
    });
    if (!res.ok) return { startM: null, endM: null };
    const data = (await res.json()) as {
      points?: { elevM?: number | null }[];
    };
    const pts = data.points ?? [];
    const start = pts[0]?.elevM;
    const end = pts[pts.length - 1]?.elevM;
    return {
      startM: start != null && Number.isFinite(start) ? start : null,
      endM: end != null && Number.isFinite(end) ? end : null,
    };
  } catch {
    return { startM: null, endM: null };
  }
}

File: mobile/lib/domain/routing/trail_access.dart
import 'dart:math' as math;

/// Trailhead orientation: gravity uses elevation (top = start), else nearest end.
class OrientedTrail {
  const OrientedTrail({
    required this.geometry,
    required this.entryLat,
    required this.entryLng,
    required this.exitLat,
    required this.exitLng,
    required this.reversed,
    required this.usedElevation,
  });

  /// [lng, lat] pairs, downhill / ride direction.
  final List<List<double>> geometry;
  final double entryLat;
  final double entryLng;
  final double exitLat;
  final double exitLng;
  final bool reversed;
  final bool usedElevation;
}

/// Minimum drop before elevation wins over GPS-nearest.
const double kTrailElevDecideMinM = 8;

double trailAccessHaversineKm(
  double lat1,
  double lng1,
  double lat2,
  double lng2,
) {
  const r = 6371.0;
  final dLat = _rad(lat2 - lat1);
  final dLng = _rad(lng2 - lng1);
  final a = math.sin(dLat / 2) * math.sin(dLat / 2) +
      math.cos(_rad(lat1)) *
          math.cos(_rad(lat2)) *
          math.sin(dLng / 2) *
          math.sin(dLng / 2);
  return 2 * r * math.asin(math.min(1.0, math.sqrt(a)));
}

double _rad(double deg) => deg * math.pi / 180;

OrientedTrail orientTrail({
  required List<List<double>> geometry,
  required double fromLat,
  required double fromLng,
  double? startElevM,
  double? endElevM,
  required bool preferDownhill,
}) {
  if (geometry.length < 2) {
    final p = geometry.isEmpty ? <double>[0, 0] : geometry.first;
    return OrientedTrail(
      geometry: geometry,
      entryLat: p.length > 1 ? p[1] : fromLat,
      entryLng: p.isNotEmpty ? p[0] : fromLng,
      exitLat: p.length > 1 ? p[1] : fromLat,
      exitLng: p.isNotEmpty ? p[0] : fromLng,
      reversed: false,
      usedElevation: false,
    );
  }

  final first = geometry.first;
  final last = geometry.last;
  final startLng = first[0];
  final startLat = first[1];
  final endLng = last[0];
  final endLat = last[1];

  var reverse = false;
  var usedElevation = false;

  final startE = startElevM;
  final endE = endElevM;
  if (preferDownhill &&
      startE != null &&
      endE != null &&
      startE.isFinite &&
      endE.isFinite &&
      (startE - endE).abs() >= kTrailElevDecideMinM) {
    usedElevation = true;
    reverse = startE < endE;
  } else {
    final dFirst = trailAccessHaversineKm(fromLat, fromLng, startLat, startLng);
    final dLast = trailAccessHaversineKm(fromLat, fromLng, endLat, endLng);
    reverse = dLast < dFirst;
  }

  final geo = reverse ? geometry.reversed.toList() : geometry;
  return OrientedTrail(
    geometry: geo,
    entryLat: geo.first[1],
    entryLng: geo.first[0],
    exitLat: geo.last[1],
    exitLng: geo.last[0],
    reversed: reverse,
    usedElevation: usedElevation,
  );
}
